### PR TITLE
ci: Prepare for shard-based unit testing

### LINKED ISSUE #14572: Use Jest `--shard` option to speed up tests
### What would you like Renovate to be able to do?

Jest `v28` comes with the `--shard` option, which splits the testing load over multiple machines. We could probably use this to speed up our Jest tests.

### If you have any ideas on how this should be implemented, please tell us here.

Try out the `--shard` option in Jest [v28.0.0](https://github.com/facebook/jest/releases/tag/v28.0.0) and see if it's actually faster.

Here are the relevant docs:

- [Jest docs for `--shard`](https://jestjs.io/docs/cli#--shard)
- [Jest `v28` blog, sharding of test runs](https://jestjs.io/blog/2022/04/25/jest-28#sharding-of-test-run)

Now that we're using Jest `v28` we should:

- [ ] Test if using `--shard` is faster
- [ ] Edit GitHub Action configuration files as needed for `--shard` to work

### Is this a feature you are interested in implementing yourself?

No

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

- Implements necessary sharding code inside `jest.config.ts`
- Extends `setup` job for outputting shard array

## Context

- Closes: #14572

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [x] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by viceice — COMMENTED** [review]
please add some tsdoc's to clarify what's happening. the code looks pretty complicated

File: jest.config.ts
Comment (HonkingGoose) [inline jest.config.ts:40]: ### Bug with `--shard` and `coverageThreshold` happens here maybe?

Did you check that the test coverage reports are still good, and that they check if everything is covered? As far as I understand there's a bug when you run Jest with `--shard` _and_ a `coverageThreshold` set:

-  https://github.com/facebook/jest/issues/12751

Maybe the same, or similar bug happens here?

File: jest.config.ts
Comment (HonkingGoose) [inline jest.config.ts:63]: ### Are we sure that all sharded tests have run?

It looks like you're manually setting a `shardMap`? I have some questions:

- How do we know that we have ran all the tests at the end?
- What happens if a directory which should be tested is _not_ on that `shardMap`?

File: jest.config.ts
Comment (HonkingGoose) [inline jest.config.ts:50]: ### What about our own issue?

Does your `TEST_SHARD` thing help (or close) this issue?

- https://github.com/renovatebot/renovate/issues/14572

Comment (zharinov) [reply jest.config.ts:50]: I didn't even try because the most recent comment in this issue starts with 'This is a big problem for us'. Instead, we can achieve the same effect by specifying shards manually and generating a couple of Jest config fields from them.

Comment (HonkingGoose) [reply jest.config.ts:50]: > I didn't even try because the most recent comment in this issue starts with 'This is a big problem for us'.

Yeah, trying `jest --shard` is probably a bad idea, seeing as the coverage collection bug has not been fixed (yet). So I understand you skipping over that issue thread. 😄 

> Instead, we can achieve the same effect by specifying shards manually and generating a couple of Jest config fields from them.

This sounds really cool and clever! Do you think we should close #14572 with your PR then?

Comment (zharinov) [reply jest.config.ts:50]: > This sounds really cool and clever! 

Thank you 😊

> Do you think we should close https://github.com/renovatebot/renovate/issues/14572 with your PR then?

Yes, the PR title already updated

File: jest.config.ts
Comment (zharinov) [reply jest.config.ts]: Each shard item includes everything that matches its pattern, except everything from previously defined shards. It means, `other` should test everything that wasn't defined earlier.

File: jest.config.ts
Comment (zharinov) [reply jest.config.ts:43]: With our approach, we're basically apply *chroot* logic to the tests run.

The only problem with threshold I see is that cross-shard coverage doesn't count anymore, so we have to write *fair* tests, e.g. test `versioning` files inside `versioning` folder. This way, for example, if manager tests call versioning API, it won't improve versioning coverage anymore.

It's the side effect, but from code quality perspective, it actually looks good to me.
But it means we have to lower our threshold values per-shard and pay this technical debt eventually.

Comment (HonkingGoose) [reply jest.config.ts:43]: @rarkins what do you think? Reducing the test threshold seems a big change to me. We should be careful that keep test coverage high, even if the coverage system can't enforce it automatically.

I like the idea of "fair tests". I'll let the maintainers chime in if this "fair test" thing is do-able.

**Comment by viceice** [comment]
> I plan to create GitHub actions `matrix` dynamically by using `jq` utility right before `build` job run. For this, I need to store shard config separately.

but we need a fixed number of named actions jobs, to make them mandatory.

**Comment by JamieMagee** [comment]
Is it going to be a manual task to update the thresholds?

**Comment by zharinov** [comment]
> Is it going to be a manual task to update the thresholds?

I'll figure it out little bit later

File: jest.config.ts
Comment (viceice) [inline jest.config.ts:153]: please no nested functions

**Comment by zharinov** [comment]
The last step of this PR is to query github `compare` endpoint in the `setup` job, get the list of changed files and detect affected shards

**Comment by zharinov** [comment]
This steps implements test sharding, next step is to enable it

**Comment by viceice** [comment]
> The last step of this PR is to query github `compare` endpoint in the `setup` job, get the list of changed files and detect affected shards

🤔 the merge queue should always run full test suite to be safe. is that the plan?

File: .github/test-shards.py
Comment (viceice) [inline .github/test-shards.py:1]: why python and not Javascript?

File: jest.config.ts
Comment (viceice) [inline jest.config.ts:4]: whoops, that import is wrong, can break in future. but can be fixed in separate PR.

**Comment by zharinov** [comment]
> 🤔 the merge queue should always run full test suite to be safe. is that the plan?

Yes, I actually think the only case for partial test run is draft PRs

**Review by viceice — COMMENTED** [review]
we need to fix coverage to get full coverage on all shards

**Comment by zharinov** [comment]
> we need to fix coverage to get full coverage on all shards

Unfortunately, it can take a while. Usually the coverage obtained from random parts of codebase is hard to fix.

File: .github/test-shards.py
Comment (zharinov) [reply .github/test-shards.py:1]: Otherwise, we need to run `setup-node` and, in some cases, await for `yarn install`, while `python` is ready to run

File: jest.config.ts
Comment (zharinov) [reply jest.config.ts:4]: Do you mean `node:os` and `node:v8`too?

File: .github/test-shards.py
Comment (zharinov) [reply .github/test-shards.py:1]: Actually, I'm not sure, node should be available by default too

Comment (zharinov) [reply .github/test-shards.py:1]: I think we actually should use our `setup-node` action to make sure the cache is populated, otherwise lots of small actions will run `yarn install` on each cache miss. Running `yarn install` in the `setup` job isn't 100% guarantee, but in 99% cases it will work well I think. So yeah, I have a nice code rewriting task for Copilot chat.

**Comment by zharinov** [comment]
- Waiting for: #22686

**Review by viceice — APPROVED** [review]

**Pull request merged**
